pilot/pkg/networking/core/v1alpha3/tracing.go

```diff
@@ -0,0 +1,480 @@
+	configureTracingFromSpec(spec, opts, hcm)
+}
+
+func configureTracingFromSpec(spec *telemetrypb.Telemetry, opts buildListenerOpts, hcm *hpb.HttpConnectionManager) {
+	meshCfg := opts.push.Mesh
+	proxyCfg := opts.proxy.Metadata.ProxyConfigOrDefault(opts.push.Mesh.DefaultConfig)
+
+	if len(spec.GetTracing()) == 0 {
+		if !meshCfg.EnableTracing {
+			log.Debugf("no valid tracing configuration found")
+			return
+		}
+		// use the prior configuration bits of sampling and custom tags
+		hcm.Tracing = &hpb.HttpConnectionManager_Tracing{}
+		configureSampling(hcm.Tracing, 0.0, proxyCfg)
+		configureCustomTags(hcm.Tracing, map[string]*telemetrypb.Tracing_CustomTag{}, proxyCfg)
+		if proxyCfg.GetTracing().GetMaxPathTagLength() != 0 {
+			hcm.Tracing.MaxPathTagLength = wrapperspb.UInt32(proxyCfg.GetTracing().MaxPathTagLength)
+		}
+		return
+	}
+
+	if len(spec.Tracing) > 1 {
+		log.Warnf("invalid number of tracing configurations provided; using first configuration found")
+	}
+
+	tracingCfg := spec.Tracing[0]
+
+	if tracingCfg.DisableSpanReporting {
+		return
+	}
+
+	// provider config
+	providerName := meshCfg.GetDefaultProviders().GetTracing()
+	if len(tracingCfg.Providers) > 0 {
+		// only one provider is currently supported, safe to take first
+		providerName = tracingCfg.Providers[0].Name
+	}
+
+	plat, ok := opts.proxy.Metadata.Raw["PLATFORM_METADATA"].(map[string]interface{})
+	if !ok {
+		log.Warnf("could not get platform metadata: raw meta = %#v", opts.proxy.Metadata.Raw)
+		log.Warnf("metadata: %#v", opts.proxy.Metadata)
+		plat = map[string]interface{}{}
+	}
+
+	providerConfigured := false
+	for _, p := range meshCfg.ExtensionProviders {
+		if strings.EqualFold(p.Name, providerName) {
+			tcfg, err := configureFromProviderConfig(opts.push, plat, p)
+			if err != nil {
+				log.Warnf("not able to configure requested tracing provider %q: %v", p.Name, err)
+				continue
+			}
+			hcm.Tracing = tcfg
+			providerConfigured = true
+			break
+		}
+	}
+
+	if !providerConfigured {
+		log.Warnf("No provider was configured for tracing")
+		hcm.Tracing = &hpb.HttpConnectionManager_Tracing{}
+		// TODO: transition to configuring providers from proxy config here?
+		// something like: configureFromProxyConfig(tracingCfg, opts.proxy.Metadata.ProxyConfig.Tracing)
+	}
+
+	// gracefully fallback to MeshConfig configuration. It will act as an implicit
+	// parent configuration during transition period.
+	configureSampling(hcm.Tracing, tracingCfg.RandomSamplingPercentage, proxyCfg)
+	configureCustomTags(hcm.Tracing, tracingCfg.CustomTags, proxyCfg)
+
+	// if there ia configured max tag length somewhere, fallback to it.
+	if hcm.GetTracing().GetMaxPathTagLength() == nil && proxyCfg.GetTracing().GetMaxPathTagLength() != 0 {
+		hcm.Tracing.MaxPathTagLength = wrapperspb.UInt32(proxyCfg.GetTracing().MaxPathTagLength)
+	}
+}
+
+// TODO: follow-on work to enable bootstrapping of clusters for $(HOST_IP):PORT addresses.
```

Comment:
does this need envoy support?

Reply:
Or i guess we can set it in metadata and pass it back. That is horrible for caching, but realistically we are a long ways  from caching LDS

Reply:
I'm not sure what the best approach here is. We ultimately want a cluster created with the address, like we do in the bootstrap currently, but without having to resort to bootstrap.